src/esp/gfx_batch/Hbao.cpp

```diff
@@ -0,0 +1,1069 @@
+ private:
+  enum : Mn::Int { LinearDepthTextureBinding = 0 };
+
+ public:
+  explicit ViewNormalShader() {
+    Cr::Utility::Resource rs{"gfx-batch-shaders"};
+
+    // TODO use our own
+    Mn::GL::Shader vert{GlslVersion, Mn::GL::Shader::Type::Vertex};
+    vert.addSource(rs.getString("hbao/fullscreenquad.vert"));
+
+    Mn::GL::Shader frag{GlslVersion, Mn::GL::Shader::Type::Fragment};
+    frag.addSource(rs.getString("hbao/viewnormal.frag"));
+
+    CORRADE_INTERNAL_ASSERT(vert.compile());
+    CORRADE_INTERNAL_ASSERT(frag.compile());
+
+    attachShaders({vert, frag});
+    CORRADE_INTERNAL_ASSERT(link());
+
+    projectionInfoUniform_ = uniformLocation("projInfo");
+    projectionOrthographicUniform_ = uniformLocation("projOrtho");
+    inverseFullResolutionUniform_ = uniformLocation("InvFullResolution");
+    setUniform(uniformLocation("texLinearDepth"), LinearDepthTextureBinding);
+  }
+
+  ViewNormalShader& setProjectionInfo(const Mn::Vector4& info) {
+    setUniform(projectionInfoUniform_, info);
+    return *this;
+  }
+
+  ViewNormalShader& setProjectionOrthographic(Mn::Int orthographic) {
+    setUniform(projectionOrthographicUniform_, orthographic);
+    return *this;
+  }
+
+  ViewNormalShader& setInverseFullResolution(const Mn::Vector2& resolution) {
+    setUniform(inverseFullResolutionUniform_, resolution);
+    return *this;
+  }
+
+  ViewNormalShader& bindLinearDepthTexture(Mn::GL::Texture2D& texture) {
+    texture.bind(LinearDepthTextureBinding);
+    return *this;
+  }
+
+ private:
+  Mn::Int projectionInfoUniform_, projectionOrthographicUniform_,
+      inverseFullResolutionUniform_;
+};
+
+class HbaoCalcShader : public Mn::GL::AbstractShaderProgram {
+ private:
+  enum : Mn::Int {
+    UniformBufferBinding = 0,
+    LinearDepthTextureBinding = 0,
+    ViewNormalTextureBinding = 1,
+    RandomTextureBinding = 1,
+    OutputImageBinding = 0
+  };
+
+ public:
+  enum Layered { Off = 0, ImageLoadStore = 1, GeometryShader = 2 };
+
+  explicit HbaoCalcShader(Mn::NoCreateT)
+      : Mn::GL::AbstractShaderProgram{Mn::NoCreate} {}
+
+  explicit HbaoCalcShader(bool deinterleaved,
+                          bool blur,
+                          Layered layered,
+                          bool textureArrayLayer)
+      : deinterleaved_{deinterleaved},
+#ifndef MAGNUM_TARGET_WEBGL
+        blur_{blur},
+#endif
+        textureArrayLayer_{textureArrayLayer},
+        layered_{layered} {
+    CORRADE_INTERNAL_ASSERT(deinterleaved || layered == Layered::Off);
+    CORRADE_INTERNAL_ASSERT(deinterleaved || !textureArrayLayer);
+
+    Cr::Utility::Resource rs{"gfx-batch-shaders"};
+
+    // TODO use our own
+    Mn::GL::Shader vert{GlslVersion, Mn::GL::Shader::Type::Vertex};
+    vert.addSource(rs.getString("hbao/fullscreenquad.vert"));
+
+#ifndef MAGNUM_TARGET_WEBGL
+    const bool passthroughSupported =
+/* Needs
+ * https://github.com/mosra/magnum/commit/5f287df332fee7fcc2df6e8055853b86d1db3ed2
+ */
```

Comment:
Does our current Magnum version support this?

Reply:
No, and that's why I put that comment here in the first place :) I need to make a Magnum update PR first, but I'm delaying that to once I'm back in the project as the set of changes since last update is rather nontrivial.

Same case is for the near/far plane extraction.